File: Losser/LINFLosser.jl
module LINFLosserModule
include("./Losser.jl")
import .LosserModule: computeLoss, Homography
import LinearAlgebra

#Extends the computeLoss function from LosserModule
function computeLoss(H_true::Homography, H_computed::Homography)::Float64
    img_l_inf_true = H_true.H[3, :]
    img_l_inf_true = img_l_inf_true / img_l_inf_true[3]

    img_l_inf_computed = H_computed.H[3, :]
    img_l_inf_computed = img_l_inf_computed / img_l_inf_computed[3]

    #Return the angle between the two lines by computing the dot product
    return LinearAlgebra.dot(img_l_inf_true, img_l_inf_computed) / (LinearAlgebra.norm(img_l_inf_true) * LinearAlgebra.norm(img_l_inf_computed))
end
end

File: src/HomoTopiContinuation/Losser/LinfLosser.py
from HomoTopiContinuation.Losser.Losser import Losser
from HomoTopiContinuation.DataStructures.datastructures import Homography
import numpy as np


class LinfLosser(Losser):
    """
    Class to compute the loss between two homographies using the angle between the lines at infinity.
    """

    def computeLoss(h_true: Homography, h_computed: Homography) -> float:
        """
        Compute the loss between two homographies using the angle between the lines at infinity.

        Args:
            h_true (Homography): The true homography
            h_computed (Homography): The computed homography

        Returns:
            float: The loss between the two homographies
        """
        l_inf_true = h_true()[2, :]
        l_inf_true_normal = l_inf_true[:2]

        l_inf_computed = h_computed()[2, :]
        l_inf_computed_normal = l_inf_computed[:2]

        angle = None
        if np.linalg.norm(l_inf_true_normal) > 0 and np.linalg.norm(l_inf_computed_normal
                                                                    ) > 0:
            angle = np.arccos(np.abs(l_inf_true_normal @ l_inf_computed_normal) / (np.linalg.norm(
                l_inf_true_normal) * np.linalg.norm(l_inf_computed_normal))) * 180 / np.pi

        return angle
